### ISSUE TITLE
numerical stability for logSigmoid

### ISSUE THREAD

**Issue by binxuan** [pytorch/pytorch#1835]
Hi,

I am using pytorch to implement a word2vec-like model. I frequently faced overflow/underflow with logsigmoid operation.

For example, if I run pytorch like this F.logsigmoid(autograd.Variable(torch.FloatTensor([-300.0]))) the result is -inf. It would be better if pytorch could implement the logsigmoid with overflow check.
 
Thanks

**Comment by binxuan** [comment]
Using function below seems to be a better choice.

def log_sigmoid(x):
    return torch.clamp(x, max=0)-torch.log(torch.exp(-torch.abs(x))+1)

**Comment by apaszke** [comment]
Are you sure the function you proposed is correct? For me it returns -300, but log(sigmoid(-300)) should be a very very large negative number (which is exactly why you're getting -inf). I don't think there's a lot we can do about that.

**Comment by binxuan** [comment]
log sigmoid(-300)= log(1/(1+exp(300))) which approximately equals to log exp(1/exp(300)).

**Comment by apaszke** [comment]
I'd say it's more like `log(sigmoid(-300)) ~~ log(1/(1 + exp(300)) ~~ log(1/exp(300)) ~~ log(0) ~~ -inf` where `~~` is approximately equal

**Comment by alykhantejani** [comment]
@apaszke I can add this if no one is working on it?

**Comment by apaszke** [comment]
Sure, go ahead! Thanks!

**Comment by soumith** [comment]
fixed in master via https://github.com/pytorch/pytorch/pull/2211

**Comment by aoken** [comment]
There are a method to correctly calculate.

F.logsigmoid(autograd.Variable(torch.DoubleTensor([-300.0])))

you get correct answer.
